### ISSUE TITLE
surface material index out of bounds error when instancing model loaded from another thread

### ISSUE THREAD

**Issue by afonsolage** [godotengine/godot#53654]
### Godot version

3.4.beta6

### System information

Linux Mint 20.2 Uma base: Ubuntu 20.04 focal GLS3

### Issue description

When trying to `.instance()` a `PackedScene` which was `.load()`ed in another thread, I got the following error:

```
E 0:00:00.748   instance_set_surface_material: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
  <C++ Source>  servers/visual/visual_server_scene.cpp:730 @ instance_set_surface_material()
  <Stack Trace> Main.gd:15 @ _add_node()
```

Tested with Godot 3.3, 3.3.1 and 3.3.3 and the same error happens

### Steps to reproduce

1. Open the project
2. Run the project
3. Check output

### Minimal reproduction project

Here is a minimal project to reproduce the error: [godot_test_surfaces_material.zip](https://github.com/godotengine/godot/files/7318409/godot_test_surfaces_material.zip)

The relevant part is the following:
```python
func _ready():
	thread.start(self, "_thread_function")


func _thread_function():
	var packed_scene = ResourceLoader.load("res://tree.tscn", "PackedScene", false)
	call_deferred("_add_node", packed_scene)


func _add_node(packed_scene):
	add_child(packed_scene.instance())
```

Which produces two errors, since the example model has 2 meshes:
```
E 0:00:00.748   instance_set_surface_material: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
  <C++ Source>  servers/visual/visual_server_scene.cpp:730 @ instance_set_surface_material()
  <Stack Trace> Main.gd:15 @ _add_node()
E 0:00:00.748   instance_set_surface_material: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
  <C++ Source>  servers/visual/visual_server_scene.cpp:730 @ instance_set_surface_material()
  <Stack Trace> Main.gd:15 @ _add_node()
```

If I change the code to instantiate within the thread itself, no error is shown:
```python
func _ready():
	thread.start(self, "_thread_function")


func _thread_function():
	var packed_scene = ResourceLoader.load("res://tree.tscn", "PackedScene", false)
	call_deferred("_add_instanced_node", packed_scene.instance())


func _add_instanced_node(node):
	add_child(node)
```

**Comment by Calinou** [comment]
PS: You can use `gdscript` syntax highlighting for code blocks on GitHub. I edited your post accordingly :slightly_smiling_face:

**Comment by Gromph** [comment]
We're running into these errors occasionally, a texture on a model will sometimes be black and log similar errors.

Is there any sort of work around other then don't load on a seperate thread? Is there a way to double check if everything loaded correctly before trying to use the scene?

**Comment by mojoyup** [comment]
It seems to happen when preloading scenes with MeshInstances that have both a custom mesh as well as an assigned Material. So, likely a confusion between loading the obj textures along with in editor textures, correct?

I haven't tested with in-editor shapes, only imported meshes.

**Comment by ouyangtianhan** [comment]
@mojoyup  same issue here with 3.4.3.
Reimport the mesh resolve that problem. But still some mesh's materials gone problem. Donnot know how to resolve that.
It works well when running it on the desktop godot editor, but some materials mssing when exported to mobile app.

**Comment by ouyangtianhan** [comment]
btw, I know how to fix the other materials missing problem now. For example, you have a materials named: **test.material**, if you wanna change its texture image by code, then you have to drag the images **each** to the material's `texture property` **each one time.** 
e.g.
![image](https://user-images.githubusercontent.com/1784053/168957660-43a2c580-487d-4648-a06b-84d732655433.png)

**Comment by ouyangtianhan** [comment]
So I guess there's some problem with the **editor's resource import & cache refresh** function

**Comment by ouyangtianhan** [comment]
OK guys, plz forget what I've posted.......
really cannot know why the materials or png files cannot be loaded or used by code....always show black materials when running **exported iOS project.**
I give up for now.

**Comment by Calinou** [comment]
@ouyangtianhan This is an unrelated issue. If the texture is never imported in the first place, you won't be able to `load()` it directly. Instead, you need to load the file using the File class then load its contents into an Image using the `load_png_from_buffer()` method.

**Comment by ouyangtianhan** [comment]
@Calinou  I think I have imported that png file. I tested in 2 ways:
1. load the png by `m = load("res://ball/test.png") `and then `$BallMesh.material_override.albeto_texture = m`
2. make a material resource and drag the png file into the material's albeto texture, and then change the ball's `material_override `in code by `load()` that material, like `$BallMesh.material_override = load("res://ball/test.material")`

Well, both not work when running in exported iOS project.
**BUT**, both work when running it in editor mode (desktop).

**Comment by Calinou** [comment]
> Well, both not work when running in exported iOS project.
> BUT, both work when running it in editor mode (desktop).

It doesn't work because the imported resource is not included within the exported PCK. You will surely reproduce the same issue if you export the project to a desktop platform and run it.

**Comment by ouyangtianhan** [comment]
@Calinou oh, so am a bit confused now, even though I use that png file in code, but it won't be exported into PCK, right?
If so, how can I mark it to be able to export into PCK then?
Actually I also tried: `var image = preload("res://ball/test.png")`, but that's also not working.

**Comment by ouyangtianhan** [comment]
Hi all, as I kept researching the problem and waiting for the new Godot version hoping it to resolve that problem, and finally am so excited that the new verions **3.5.1** released, and I tried it with my project. And guess what, boooooom, it does NOT resolve my problem at all [:sad:]

But 'cause I've been researching this for 1 year, I found out that reason today.........seems........

**Before:**
I used image size: **1024x512** for my mesh's materials. It worked well on Mac desktop, windows desktop. 
But not working(showing correct image) on  iOS app.
**After:**
I changed the image size to: **512x512**, boooooooom, iOS OKKKKKKKKed!

**Solution:**
well, I still not known the root working principle, but at least that's the reason and can resolve thus problem.
I should learn more the basic concept and principles other than wondering the GDScript's bug ;)

**Comment by jitspoe** [comment]
I'm also hitting this error, but it only seems to happen when adding something loaded from a thread to a viewport.  I've been using multithreaded loading (using a slightly modified script from the example on the Godot site) for my level loading and haven't gotten the issue.  I'm doing a special case thing now where I want to load a level to display on a viewport texture, and that is giving me tons of errors.  I've tried adding deferred calls and delays and such, but I haven't had any luck getting rid of the error.  It seems to happen as soon as I do an add_child() call.

Errors:

```
E 0:00:06.739   VisualServerScene::instance_set_surface_material: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
  <C++ Source>  servers\visual\visual_server_scene.cpp:1155 @ VisualServerScene::instance_set_surface_material()
  <Stack Trace> MonitorViewport.gd:64 @ LoadLevelResource()
E 0:00:06.744   RID_Owner<struct RasterizerStorageGLES3::Material>::get: Condition "!id_map.has(p_rid.get_data())" is true. Returned: nullptr
  <C++ Source>  E:\Projects\godot\godot_engine_mine_32\core/rid.h:142 @ RID_Owner<struct RasterizerStorageGLES3::Material>::get()
  <Stack Trace> MonitorViewport.gd:66 @ LoadLevelResource()
E 0:00:06.744   RasterizerStorageGLES3::material_set_param: Condition "!material" is true.
  <C++ Source>  drivers\gles3\rasterizer_storage_gles3.cpp:2575 @ RasterizerStorageGLES3::material_set_param()
  <Stack Trace> MonitorViewport.gd:66 @ LoadLevelResource()
```


Don't mind the hacky code, but maybe this will help shed some light on things:

```
extends Viewport

var LevelLoaded : Spatial
onready var CameraNode := $Camera


var LevelsWithKills := []
var PlayerKillPosition : Vector3
var PlayerKillVelocity : Vector3
var EnemyKillPosition : Vector3
var EnemyKillVelocity : Vector3

var LevelQueued := false
var QueuedPath : String

var TEMPFirstObjectPath := ""


const LOAD_DELAY := 0.5
var TimeSinceReady := 0.0
var ReadyDelayedCalled := false

func _ready():
	# This is loading within a thread, and a thread within a thread causes problems, so delay it.
	#call_deferred("ReadyDelayed")
	pass # Call with a delay


func ReadyDelayed():
	KillCam.SetMonitorViewport(self)
	# Find the level where kills happened.
	var KillCamSaveData := KillCam.KillInfoSaved
	for ObjectPath in KillCamSaveData:
		if (!TEMPFirstObjectPath):
			TEMPFirstObjectPath = ObjectPath
		var LevelName := Global.GetLevelNameFromObjectPath(ObjectPath)
		if (!LevelsWithKills.has(LevelName)):
			LevelsWithKills.append(LevelName)
	if (LevelsWithKills.size() > 0):
		LoadLevelThreaded(LevelsWithKills[0]) # TOMAYBEDO: Might need to delay this a couple seconds to avoid issues...


func _physics_process(delta : float):
	TimeSinceReady += delta
	if (!ReadyDelayedCalled && TimeSinceReady > LOAD_DELAY):
		ReadyDelayed()
		ReadyDelayedCalled = true
	# PlayerSpawner.gd sets camera position to spawn location.
	#CameraNode.global_transform.origin.x += delta * rand_range(-4, 4);
	CameraNode.global_transform.origin = PlayerKillPosition
	CameraNode.global_transform.origin.z = 40.0
	if (LevelQueued):
		if (resource_queue.is_ready(QueuedPath)):
			LevelQueued = false
			if (OS.is_debug_build()):
				print("Monitor level ready!")
			call_deferred("LoadLevelResource")
		else:
			if (OS.is_debug_build()):
				print("Monitor level loading.")


func LoadLevelResource():
	LevelLoaded = resource_queue.get_resource(QueuedPath).instance()
	add_child(LevelLoaded)
	call_deferred("SetFocusTarget")


func SetFocusTarget():
	KillCam.SetCurrentFocusTarget(TEMPFirstObjectPath) # TODO: Cycle through objects and levels


func LoadLevelThreaded(LevelName : String):
	QueuedPath = "res://Levels/" + LevelName + ".tscn"
	LevelQueued = true
	resource_queue.queue_resource(QueuedPath)
```

And here's the resource queue code:
```
extends Node # Doesn't seem to work as a singleton if it doesn't extend anything.

var thread : Thread
var mutex : Mutex
var sem : Semaphore
var exit_thread := false


const USE_MULTITHREADING := true


var queue := []
var pending := {}


func _lock(_caller):
	mutex.lock()


func _unlock(_caller):
	mutex.unlock()


func _post(_caller):
	sem.post()


func _wait(_caller):
	sem.wait()


func _init():
	#print("resource_queue init.")
	set_process(false) # Don't process until we've queued something to load (only happens in non-threaded version)
	if (USE_MULTITHREADING):
		start_threaded()


# warning-ignore:unused_argument
# Single-threaded loading process.  Load one resource at a time.
func _process(delta : float):
	_process_queue_single()

func _process_queue_single():
	if (queue.size() > 0):
		var res : ResourceInteractiveLoader = queue[0]
		var ret := res.poll()
		if (ret != OK):
			var path = res.get_meta("path")
			if path in pending: # else it was already retrieved
				pending[path] = res.get_resource()
			queue.erase(res) # something might have been put at the front of the queue while we polled, so use erase instead of remove
	else:
		set_process(false) # Nothing left in queue.  Stop processing.


func queue_resource(path, p_in_front = false):
	if (USE_MULTITHREADING):
		_lock("queue_resource")
	else:
		set_process(true)

	if path in pending:
		if (USE_MULTITHREADING):
			_unlock("queue_resource")
		return

	elif ResourceLoader.has_cached(path):
		var res = ResourceLoader.load(path)
		pending[path] = res
		if (USE_MULTITHREADING):
			_unlock("queue_resource")
		return
	else:
		var res = ResourceLoader.load_interactive(path)
		res.set_meta("path", path)
		if p_in_front:
			queue.insert(0, res)
		else:
			queue.push_back(res)
		pending[path] = res
		if (USE_MULTITHREADING):
			_post("queue_resource")
			_unlock("queue_resource")
		return


func cancel_resource(path):
	if (USE_MULTITHREADING):
		_lock("cancel_resource")
	if path in pending:
		if pending[path] is ResourceInteractiveLoader:
			queue.erase(pending[path])
		pending.erase(path)
	if (USE_MULTITHREADING):
		_unlock("cancel_resource")


func get_progress(path):
	if (USE_MULTITHREADING):
		_lock("get_progress")
	var ret = -1.0
	if path in pending:
		if pending[path] is ResourceInteractiveLoader:
			ret = float(pending[path].get_stage()) / float(pending[path].get_stage_count())
		else:
			ret = 1.0
	if (USE_MULTITHREADING):
		_unlock("get_progress")
	return ret


func is_ready(path):
	var ret
	if (USE_MULTITHREADING):
		_lock("is_ready")
	if path in pending:
		if (pending[path] is ResourceInteractiveLoader):
			var stage = pending[path].get_stage()
			var stage_count = pending[path].get_stage_count()
			print("stage: %d / %d" % [stage, stage_count])
			ret = false
		ret = true
	else:
		ret = false

	if (USE_MULTITHREADING):
		_unlock("is_ready")

	return ret


func _wait_for_resource(res, path):
	_unlock("wait_for_resource")
	while true:
		VisualServer.sync()
		OS.delay_usec(16000) # wait 1 frame
		_lock("wait_for_resource")
		if queue.size() == 0 || queue[0] != res:
			return pending[path]
		_unlock("wait_for_resource")


func get_resource(path):
	if USE_MULTITHREADING:
		_lock("get_resource")
	if path in pending:
		if pending[path] is ResourceInteractiveLoader:
			var res = pending[path]
			if res != queue[0]:
				var pos = queue.find(res)
				queue.remove(pos)
				queue.insert(0, res)

			if (USE_MULTITHREADING):
				res = _wait_for_resource(res, path)
			else:
				while (queue.size() > 0):
					_process_queue_single()
				res = pending[path]

			pending.erase(path)
			if USE_MULTITHREADING:
				_unlock("return")
			return res

		else:
			var res = pending[path]
			pending.erase(path)
			if USE_MULTITHREADING:
				_unlock("return")
			return res
	else:
		if USE_MULTITHREADING:
			_unlock("return")
		return ResourceLoader.load(path)


func thread_process():
	_wait("thread_process")
	_lock("process")

	while queue.size() > 0:
		var res = queue[0]
		_unlock("process_poll")
		var ret = res.poll()
		_lock("process_check_queue")

		if ret == ERR_FILE_EOF || ret != OK:
			var path = res.get_meta("path")
			if path in pending: # else it was already retrieved
				pending[res.get_meta("path")] = res.get_resource()

			queue.erase(res) # something might have been put at the front of the queue while we polled, so use erase instead of remove

	_unlock("process")


func thread_func(_u):
	while true:
		mutex.lock()
		var should_exit = exit_thread # Protect with Mutex.
		mutex.unlock()

		if should_exit:
			break
		thread_process()


func start_threaded():
	mutex = Mutex.new()
	sem = Semaphore.new()
	thread = Thread.new()
	thread.start(self, "thread_func", 0)


func _exit_tree():
	mutex.lock()
	exit_thread = true # Protect with Mutex.
	mutex.unlock()

	# Unblock by posting.
	sem.post()

	# Wait until it exits.
	thread.wait_to_finish()
```

Also worth noting that the errors are inconsistent, especially if I'm debugging.  Stepping through, I usually don't get the errors, so it seems heavily dependent on the multithreading.

**Comment by wagfeliz** [comment]
This error is happening on 4.03 stable on linux as well, not really related with loading images or materials by code.
In my case, I have a mesh created with ImmediateMesh, and I am just trying to change the color of the mesh : 

```
func changeMaterialColor(c:Color)->bool:
	arrowmesh.get_active_material(0).albedo_color = c
	return true
```
Let me know if any needs the intire code, its shared in the forum, but its needs some update to reflect this bug.

**Comment by wagfeliz** [comment]
In my case, it looks wrong : 

```
Ref<Material> MeshInstance3D::get_active_material(int p_surface) const {
	Ref<Material> mat_override = get_material_override();
	if (mat_override.is_valid()) {
		return mat_override;
	}

	Ref<Material> surface_material = get_surface_override_material(p_surface);
	if (surface_material.is_valid()) {
		return surface_material;
	}

	Ref<Mesh> m = get_mesh();
	if (m.is_valid()) {
		return m->surface_get_material(p_surface);
	}

	return Ref<Material>();
```
Its trying to return get_surface_override_material(p_surface) when there is no override material, and it gets the error in my case. It should check if there is any override material first...

**Comment by Ahopness** [comment]
also happened to me, im on 3.5.2.stable running on windows64x
its fine when i load my scene using the Threads class, but if i add a Sprite3D to the scene, it gives me that error and doesnt render

this is the code im using to load the scene:
```gdscript
signal scene_loaded
var scene_loading_thread
func change_scene_async(scene_path :String, show_loading_icon :bool = true, anm_vel :float = 1, anm_name :String = "fade"):
	$Transitions/anm.play(anm_name, -1, anm_vel, false)
	yield($Transitions/anm, "animation_finished")
	if show_loading_icon: 
		toggle_loading_icon(false, true)
		yield($Iconography/loading/tween, "tween_completed")
	
	scene_loading_thread = Thread.new()
	scene_loading_thread.start(self, "_change_scene", scene_path, 2)
	yield(self, "scene_loaded")
	yield(get_tree(), "idle_frame")
	yield(get_tree(), "idle_frame")
	
	if show_loading_icon: 
		toggle_loading_icon(false, false)
		yield($Iconography/loading/tween, "tween_completed")
	$Transitions/anm.play(anm_name, -1, -anm_vel, true)
func _change_scene(_scene_path :String):
	get_tree().change_scene(_scene_path)
	emit_signal("scene_loaded")
func _exit_tree():
	if scene_loading_thread != null:
		scene_loading_thread.wait_to_finish()
```

error that debugs _twice_ when the scene finishes loading:
```cpp
E 0:00:05.048   instance_set_surface_material: Index p_surface = 0 is out of bounds (instance->materials.size() = 0).
  <C++ Source>  servers/visual/visual_server_scene.cpp:1139 @ instance_set_surface_material()
```

it works fine if i load the scene from the editor using the "play scene (f6)" button

**Comment by jitspoe** [comment]
So, while the 0.5s delay usually works for me, I did run into some cases where it didn't.  Every once in a while my computer gets in this state where it loads things off of the hard drive very slowly, and in that case, I started getting the errors again.  The delay does not 100% solve it, but it solves it most of the time, unless the load times are too slow.  It's almost like not all the sub-resources are getting loaded before is_ready() returns true?

**Comment by KYRYAMBA** [comment]
For my task, I solved this problem by setting material in GeometryInstance -> material_override. 
Error and black material occurs when i was trying to set in into "material/0".
![image](https://github.com/godotengine/godot/assets/99164888/56d91bc6-59d9-4421-8ab6-5b25b407a879)
